### PR TITLE
[SDA-4612] Download rosa option

### PR DESCRIPTION
**Pull request by gdbranco**
Related issue: https://issues.redhat.com/browse/SDA-4612

# What
Add option to download latest rosa client through rosa itself

# Why
Facilitates user flow getting latest release

# Changes
`rosa download rosa`
Downloads the latest release with the following naming convention `rosa-<platform>-amd64`, amd64 is the only supported arch atm, when m1 arch is supported we need to improve this functionality.

`rosa verify rosa`
Checks if current CLI is up to date

### CODE REVIEW

File: cmd/verify/rosa/cmd.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
```
Comment (gdbranco) [inline cmd/verify/rosa/cmd.go:2]: 22

File: cmd/download/rosa/cmd.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
```
Comment (gdbranco) [inline cmd/download/rosa/cmd.go:2]: 22

File: cmd/download/rosa/cmd.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+// Get the platform name used on the oc tarball filename
+func getPlatform() string {
+	if runtime.GOOS == "darwin" {
+		return "macosx"
+	}
+	return runtime.GOOS
+}
+
+// Get the extension used for the compressed oc file
+func getExtension() string {
```
Comment (oriAdler) [inline cmd/download/rosa/cmd.go:74]: We have a similar function in `cmd/download/oc/cmd.go`,
should we extract it to use in both commands?

File: cmd/download/rosa/cmd.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	if runtime.GOOS == "windows" {
+		return "zip"
+	}
+	return "tar.gz"
+}
+
+// download will download a url to a local file. It's efficient because it will
+// write as it downloads and not load the whole file into memory. We pass an io.TeeReader
+// into Copy() to report progress on the download.
+func download(url string, filename string) error {
```
Comment (oriAdler) [inline cmd/download/rosa/cmd.go:84]: What is the difference between this function and the function in `cmd/download/oc/cmd.go`, can we re-use the function?

File: cmd/verify/rosa/cmd.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+func run(_ *cobra.Command, _ []string) {
+	rprtr := reporter.CreateReporterOrExit()
+	resp, err := http.Get(BaseReleasesFolder)
+	if err != nil {
+		rprtr.Errorf("Error setting up request for latest released rosa cli: %v", err)
+		os.Exit(1)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
```
Comment (oriAdler) [inline cmd/verify/rosa/cmd.go:55]: NP
```suggestion
	if resp.StatusCode < http.StatusOK || resp.StatusCode > 299 {
```

File: cmd/verify/rosa/cmd.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		rprtr.Errorf("Error while requesting latest released rosa cli: %d %s", resp.StatusCode, resp.Status)
+		os.Exit(1)
+	}
+	doc, err := goquery.NewDocumentFromReader(resp.Body)
+	if err != nil {
+		rprtr.Errorf("Error parsing response body: %v", err)
+		os.Exit(1)
+	}
+	possibleVersions := []string{}
```
Comment (oriAdler) [inline cmd/verify/rosa/cmd.go:64]: NP
We can decompose conditional lines 64-75 to a new function **findPossibleVersions**?

File: pkg/helper/helpers.go [hunk 7]
```diff
@@ -96,13 +96,7 @@ func SliceToMap(s []string) map[string]bool {
 }
 
 func SliceToString(s []string) string {
```
Comment (oriAdler) [inline pkg/helper/helpers.go:98]: Should it be `SliceToSortedString`?

Comment (gdbranco) [reply pkg/helper/helpers.go]: That shouldn't be in this PR :)

Comment (gdbranco) [reply pkg/helper/helpers.go]: #994 PTAL

File: cmd/download/oc/cmd.go [hunk 8]
```diff
@@ -72,73 +69,3 @@ func getPlatform() string {
 	}
 	return runtime.GOOS
 }
-
```
Comment (gdbranco) [inline cmd/download/oc/cmd.go:75]: Ended up extracting this into it's own command `verify rosa` for some reason prow wasn't able to reach mirror to check it from here which ended up breaking merge flow. Any ideas why? Can we continue like this?

File: pkg/helper/download/download.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+func GetExtension() string {
+	if runtime.GOOS == "windows" {
+		return "zip"
+	}
+	return "tar.gz"
+}
+
+// WriteCounter counts the number of bytes written to it. It implements to the io.Writer interface
+// and we can pass this into io.TeeReader() which will report progress on each write cycle.
+type WriteCounter struct {
```
Comment (oriAdler) [inline pkg/helper/download/download.go:63]: NP
We can have this new struct in the helper files since it can be used in other commands in the future.

**Pull request merged**